README.md
# SVN DiffTo 

* right click on project in workspace and choose "SVN DiffTo"
* type svn tag (default ^/trunk)
* wait to generate

## Instalation

* download latest release
* install by command: code --install-extension svndiffto-*.vsix

## Features

* shows svn diff between local project and any svn tag within same repository

## Requirements

* Ubuntu 20.04

## Release Notes

### 0.0.7

* SVN item type detection fix
* Added unknown type
* Differentiate loading type
* Generating diffto parsing with progress


CHANGELOG.md
# Change Log

## [0.0.7]

- SVN item type detection fix
- Added unknown type
- Differentiate loading type
- Generating diffto parsing with progress

## [0.0.6]

- Refactored DiffTo item type
- Added DiffTo loading item icon

## [0.0.5]

- Fix DiffTo with non-root directory and single file
- Input with value suggestion
- Input with path relative to chosen DiffTo element

## [0.0.4]

- DiffTo content icons
- DiffTo content display only filename, not full path
- Showing progress instead DiffTo content
- Dot excluded from DiffTo content
- Large letter margers

## [0.0.3]

- Instead of new tab with webview for file changes display explorer view

## [0.0.2]

- Running of command moved from command palete to project right click menu

## [0.0.1]

- Initial release

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { SVNDiffToProvider, SVNDiffToItemType } from './SVNDiffToProvider';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('svndiffto.svndiffto', (params) => {
		let workspaceFolders = vscode.workspace.workspaceFolders;
		if(typeof workspaceFolders === 'undefined')
		{
			workspaceFolders = [];
		}
		if(workspaceFolders.length === 0)
		{
			vscode.window.showErrorMessage('No workspace folders');
			return;
		}
		const workspaceFolderPaths = workspaceFolders.map(wsf => wsf.uri.path);

		const workspaceFolderPath = params.path;

		let placeHolderSufix = '';
		for(let i=0; i<workspaceFolderPaths.length; i++)
		{
			const wsfpath = workspaceFolderPaths[i];
			if(wsfpath === workspaceFolderPath)
			{
				break;
			}
			else if(workspaceFolderPath.startsWith(wsfpath))
			{
				placeHolderSufix = workspaceFolderPath.substring(wsfpath.length);
				break;
			}
		}

		const pickedProject = workspaceFolderPath.substring(workspaceFolderPath.lastIndexOf("/")+1);

		let generatingStatusMessage = vscode.window.setStatusBarMessage('fetching svn info');

		const cp1 = require('child_process');
		cp1.exec('svn info '+workspaceFolderPath, (err: string, stdout: string, stderr: string) => {
			if (err) {
				console.log('error: ' + err);
				vscode.window.showErrorMessage('error: ' + err);
				return;
			}

			generatingStatusMessage.dispose();
			generatingStatusMessage = vscode.window.setStatusBarMessage('parsing fetched svn info');

			let projectSvnInfoLines = stdout.match(/[^\r\n]+/g);

			if(projectSvnInfoLines === null)
			{
				vscode.window.showErrorMessage('svn info empty');
				return;
			}

			let relativeUrlLine = '';
			let repositoryRootLine = '';
			let isSingleFileDiff = false;
			projectSvnInfoLines.forEach(line => {
				const lineTrim = line.trim();
				if(lineTrim.startsWith('Relative URL: '))
				{
					relativeUrlLine = lineTrim;
				}
				else if(lineTrim.startsWith('Repository Root: '))
				{
					repositoryRootLine = lineTrim;
				}
				else if(lineTrim.startsWith('Name: '))
				{
					isSingleFileDiff = true;
				}
			});
			if(relativeUrlLine === '')
			{
				vscode.window.showErrorMessage('svn info no relative url line');
				return;
			}
			if(repositoryRootLine === '')
			{
				vscode.window.showErrorMessage('svn info no repository root line');
				return;
			}

			const projectRelativeUrl = relativeUrlLine.substring(14);
			const projectRepositoryRoot = repositoryRootLine.substring(17);

			const projectRelativeUrlClean = projectRelativeUrl.replace(/^(\^\/)/,"");

			generatingStatusMessage.dispose();

			const initialValue = "^/trunk" + placeHolderSufix;

			let svntagOptions: vscode.InputBoxOptions = {
				prompt: "Choose with what to diff: ",
				value: initialValue,
				placeHolder: initialValue
			};

			generatingStatusMessage.dispose();

			vscode.window.showInputBox(svntagOptions).then(value => {	
				let answer = value;
				if (!answer || typeof answer === 'undefined')
				{
					answer = initialValue;
				}
				const finalAnswer = answer;

				const answerClean = finalAnswer.replace(/^(\^\/)/,"");

				const svnOld = projectRepositoryRoot + '/' + answerClean;
				const svnNew = projectRepositoryRoot + '/' + projectRelativeUrlClean;

				generatingStatusMessage = vscode.window.setStatusBarMessage('Generating DiffTo - fetching from server...');
				vscode.window.registerTreeDataProvider(
					'svndiffto',
					new SVNDiffToProvider([
						{
							'path': 'Generating DiffTo - fetching from server...',
							'type': SVNDiffToItemType.loading
						}
					])
				);

				const cp2 = require('child_process');
				cp2.exec('svn diff --old='+svnOld+' --new='+svnNew, {maxBuffer: 1024*1024*1024}, (err: string, stdout: string, stderr: string) => {
					if (err) {
						console.log('error: ' + err);
						vscode.window.showErrorMessage('error: ' + err);
						return;
					}

					generatingStatusMessage.dispose();
					let parsingStatusMessage = vscode.window.setStatusBarMessage('Generating DiffTo - parsing');
					vscode.window.registerTreeDataProvider(
						'svndiffto',
						new SVNDiffToProvider([
							{
								'path': 'Generating DiffTo - parsing',
								'type': SVNDiffToItemType.loading
							}
						])
					);

					const svndiffContentArray = stdout.split('Index: ');
					const svndiffContentArrayLength = svndiffContentArray.length;

					const cpTemp1 = require('child_process');
					const cpTemp1Stdout = cpTemp1.execSync('mktemp');

					const tempFilePath = cpTemp1Stdout.toString().trim();
					var emptyTempPathUri = vscode.Uri.parse("file://" + tempFilePath);

					let treeData: any[] = [];
					let parsingCounter = 0;
					svndiffContentArray.forEach(element => {

						parsingCounter++;

						vscode.window.registerTreeDataProvider(
							'svndiffto',
							new SVNDiffToProvider([
								{
									'path': 'Generating DiffTo - parsing '+ parsingCounter + '/' + svndiffContentArrayLength,
									'type': SVNDiffToItemType.loading
								}
							])
						);

						if(element.trim().length === 0)
						{
							return;
						}

						const elementLines = element.split("\n");
						const path = elementLines[0];
						if(path === '.')
						{
							return; /// continue
						}
						
						const elementLines2 = elementLines[2];
						const elementLines3 = elementLines[3];
						
						var elementType = SVNDiffToItemType.unknown;

						if(elementLines2.includes('(revision ') && elementLines3.includes('(revision '))
						{
							elementType = SVNDiffToItemType.svnModify;
						}
						else if(elementLines2.endsWith('(nonexistent)') && elementLines3.includes('(revision '))
						{
							elementType = SVNDiffToItemType.svnAdd;
						}
						else if(elementLines3.endsWith('(nonexistent)') && elementLines2.includes('(revision '))
						{
							elementType = SVNDiffToItemType.svnDelete;
						}
						if(elementLines2 === 'Cannot display: file marked as a binary type.')
						{
							elementLines.forEach((elementLine: string) => {
								if(elementLine === 'Added: svn:mime-type')
								{
									elementType = SVNDiffToItemType.svnAdd;
								}
							});
						}

						let filePath = workspaceFolderPath;
						if(isSingleFileDiff === false)
						{
							filePath += '/' + path;
						}
						var openPath = vscode.Uri.parse("file://" + filePath); //A request file path

						let leftFile = null;
						let rightFile = null;

						if(elementType === SVNDiffToItemType.svnAdd)
						{
							leftFile = emptyTempPathUri;
							rightFile = openPath;
						}
						else
						{
							const cpTemp2 = require('child_process');
							const cpTemp2Stdout = cpTemp2.execSync('mktemp');

							const tempFilePath = cpTemp2Stdout.toString().trim();
							var tempPathUri = vscode.Uri.parse("file://" + tempFilePath);

							let svnPath = svnOld;
							if(isSingleFileDiff === false)
							{
								svnPath += '/' + path;
							}
							const cp3 = require('child_process');
							cp3.execSync('svn cat '+svnPath+' > '+tempFilePath, {maxBuffer: 1024*1024*1024});

							if(elementType === SVNDiffToItemType.svnDelete)
							{
								leftFile = tempPathUri;
								rightFile = emptyTempPathUri;
							}
							else /// m
							{
								leftFile = tempPathUri;
								rightFile = openPath;
							}
						}

						treeData.push({
							'path': path,
							'type': elementType,
							'leftFile': leftFile,
							'rightFile': rightFile
						});
					});

					vscode.window.registerTreeDataProvider(
						'svndiffto',
						new SVNDiffToProvider([
							{
								'path': 'Generating DiffTo - done',
								'type': SVNDiffToItemType.loading
							}
						])
					);

					vscode.window.registerTreeDataProvider(
						'svndiffto',
						new SVNDiffToProvider(treeData)
					);

					parsingStatusMessage.dispose();
					vscode.window.setStatusBarMessage('Generating DiffTo done', 2000);
				});
			});
		});
	});

	context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/SVNDiffToProvider.ts
import * as vscode from 'vscode';
import * as path from 'path';

export class SVNDiffToProvider implements vscode.TreeDataProvider<SVNDiffToItem> {
    constructor(private filesDiff: any) {}

    getTreeItem(element: SVNDiffToItem): vscode.TreeItem {
        return element;
    }

    getChildren(element?: SVNDiffToItem): Thenable<SVNDiffToItem[]> {
        if (element) {
            return Promise.resolve([]);
        } else {
            var resolveResult: SVNDiffToItem[] = [];
            this.filesDiff.forEach((element: any) => {
                let elementLabel = element.path;
                if(element.type !== SVNDiffToItemType.loading)
                {
                    elementLabel = element.path.replace(/^.*[\\\/]/, '');
                }
                const svnDiffToItem = new SVNDiffToItem(
                    elementLabel,
                    element.path,
                    element.type, 
                    element.leftFile, 
                    element.rightFile, 
                    vscode.TreeItemCollapsibleState.None
                );
                resolveResult.push(svnDiffToItem);
            });
            return Promise.resolve(resolveResult);
        }
    }
}

class SVNDiffToItem extends vscode.TreeItem {
    constructor(
      public readonly label: string,
      private diffPath: string,
      private type: SVNDiffToItemType,
      private leftFileUri: vscode.Uri,
      private rightFileUri: vscode.Uri,
      public readonly collapsibleState: vscode.TreeItemCollapsibleState
    ) {
        super(label, collapsibleState);

        if(this.type === SVNDiffToItemType.loading)
        {
            this.iconPath = {
                light: path.join(__filename, '..', '..', 'resources', 'loading_light.png'),
                dark: path.join(__filename, '..', '..', 'resources', 'loading_dark.png')
            };
        }
        else
        {
            this.description = `${this.type} - ${this.diffPath}`;
            this.tooltip = `${this.diffPath}`;
            this.command = {
                "title": "SVN DiffTo",
                "command": "vscode.diff",
                "arguments": [
                    leftFileUri,
                    rightFileUri,
                    "SVN DiffTO: " + label
                ]
            };

            let iconName = 'invalid.png';
            if(type === SVNDiffToItemType.svnModify)
            {
                iconName = 'dot_blue.png';
            }
            else if(type === SVNDiffToItemType.svnAdd)
            {
                iconName = 'dot_green.png';
            }
            else if(type === SVNDiffToItemType.svnDelete)
            {
                iconName = 'dot_grey.png';
            }
            this.iconPath = {
                light: path.join(__filename, '..', '..', 'resources', iconName),
                dark: path.join(__filename, '..', '..', 'resources', iconName)
            };
        }
    }
}

export enum SVNDiffToItemType {
    loading = 'L',
    svnModify = 'M',
    svnAdd = 'A',
    svnDelete = 'D',
    unknown = 'U'
}
  

extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    let disposable = vscode.commands.registerCommand('svndiffto.svndiffto', () => {
        let workspaceFolders = vscode.workspace.workspaceFolders;
        if (typeof workspaceFolders === 'undefined') {
            workspaceFolders = [];
        }
        if (workspaceFolders.length === 0) {
            vscode.window.showErrorMessage('No workspace folders');
            return;
        }
        const workspaceFoldersConst = workspaceFolders;
        const workspaceFolderNames = workspaceFolders.map(folder => folder.name);
        const quickPickOptions = workspaceFolderNames.map(label => ({ label }));
        const quickPick = vscode.window.createQuickPick();
        quickPick.items = quickPickOptions;
        quickPick.onDidChangeSelection(([{ label }]) => {
            quickPick.hide();
            const pickedProject = label;
            let index = -1;
            for (var i = 0; i < quickPickOptions.length; i += 1) {
                if (quickPickOptions[i]['label'] === pickedProject) {
                    index = i;
                    break;
                }
            }
            const workspaceFolderPath = workspaceFoldersConst[index].uri.path;
            let generatingStatusMessage = vscode.window.setStatusBarMessage('fetching svn info');
            const cp1 = require('child_process');
            cp1.exec('svn info ' + workspaceFolderPath, (err, stdout, stderr) => {
                if (err) {
                    console.log('error: ' + err);
                    vscode.window.showErrorMessage('error: ' + err);
                    return;
                }
                generatingStatusMessage.dispose();
                generatingStatusMessage = vscode.window.setStatusBarMessage('parsing fetched svn info');
                let projectSvnInfoLines = stdout.match(/[^\r\n]+/g);
                if (projectSvnInfoLines === null) {
                    vscode.window.showErrorMessage('svn info empty');
                    return;
                }
                const relativeUrlLine = projectSvnInfoLines[3];
                const repositoryRootLine = projectSvnInfoLines[4];
                const projectRelativeUrl = relativeUrlLine.substring(14);
                const projectRepositoryRoot = repositoryRootLine.substring(17);
                const projectRelativeUrlClean = projectRelativeUrl.replace(/^(\^\/)/, "");
                generatingStatusMessage.dispose();
                let svntagOptions = {
                    prompt: "Choose with what to diff: ",
                    placeHolder: "^/trunk"
                };
                generatingStatusMessage.dispose();
                vscode.window.showInputBox(svntagOptions).then(value => {
                    let answer = value;
                    if (!answer || typeof answer === 'undefined') {
                        answer = '^/trunk';
                    }
                    const finalAnswer = answer;
                    const answerClean = finalAnswer.replace(/^(\^\/)/, "");
                    const svnOld = projectRepositoryRoot + '/' + answerClean;
                    const svnNew = projectRepositoryRoot + '/' + projectRelativeUrlClean;
                    generatingStatusMessage = vscode.window.setStatusBarMessage('Generating DiffTo - fetching from server...');
                    const cp2 = require('child_process');
                    cp2.exec('svn diff --old=' + svnOld + ' --new=' + svnNew, { maxBuffer: 1024 * 1024 * 1024 }, (err, stdout, stderr) => {
                        if (err) {
                            console.log('error: ' + err);
                            vscode.window.showErrorMessage('error: ' + err);
                            return;
                        }
                        generatingStatusMessage.dispose();
                        let parsingStatusMessage = vscode.window.setStatusBarMessage('Generating DiffTo - parsing');
                        const svndiffContentArray = stdout.split('Index: ');
                        let filesDiff = [{
                                'path': 'Choose file',
                            }];
                        svndiffContentArray.forEach(element => {
                            if (element.trim().length === 0) {
                                return;
                            }
                            const elementLines = element.split("\n");
                            const path = elementLines[0];
                            var elementAddLines = 0;
                            var elementDeleteLines = 0;
                            elementLines.forEach((elementLine) => {
                                if (elementLine.startsWith("-")) {
                                    elementDeleteLines++;
                                }
                                else if (elementLine.startsWith("+")) {
                                    elementAddLines++;
                                }
                            });
                            var elemetType = 'm';
                            if (elementDeleteLines === 1) {
                                elemetType = 'a';
                            }
                            else if (elementAddLines === 1) {
                                elemetType = 'd';
                            }
                            filesDiff.push({
                                'path': path,
                                'diffContent': element,
                                'type': elemetType
                            });
                        });
                        parsingStatusMessage.dispose();
                        vscode.window.setStatusBarMessage('Generating DiffTo done', 2000);
                        //vscode.window.createTreeView('svnDiffTo', {
                        //	treeDataProvider: new SVNDiffToProvider()
                        //  });
                        //new FileExplorer(context);
                        const column = vscode.window.activeTextEditor
                            ? vscode.window.activeTextEditor.viewColumn
                            : undefined;
                        const panel = vscode.window.createWebviewPanel('svndiffto', "SVN DiffTO: " + pickedProject + ' <-> ' + finalAnswer, column || vscode.ViewColumn.One, {
                            // Enable scripts in the webview
                            enableScripts: true
                        });
                        let difItemsSelectContent = '';
                        filesDiff.forEach((element, index) => {
                            difItemsSelectContent += '<option value=' + index + '>' +
                                element.type + ' - ' + element.path
                                + '</option>';
                        });
                        const cssUri = panel.webview.asWebviewUri(vscode.Uri.joinPath(context.extensionUri, 'src', 'css', 'styles.css'));
                        panel.webview.html = getWebviewContent(difItemsSelectContent, cssUri);
                        // Handle messages from the webview
                        panel.webview.onDidReceiveMessage(message => {
                            switch (message.command) {
                                case 'onSelectChange':
                                    const filesDiffType = filesDiff[message.files_diff_index].type;
                                    const filesDiffPath = filesDiff[message.files_diff_index].path;
                                    if (filesDiffType === 'a') {
                                        const filePath = workspaceFolderPath + '/' + filesDiff[message.files_diff_index].path;
                                        var openPath = vscode.Uri.parse("file://" + filePath); //A request file path
                                        vscode.workspace.fs.stat(openPath).then(fStat => {
                                            vscode.workspace.fs.readFile(openPath).then(content => {
                                                if (typeof filesDiff[message.files_diff_index].tempPath === 'undefined') {
                                                    const cpTemp1 = require('child_process');
                                                    cpTemp1.exec('mktemp', (err, stdout, stderr) => {
                                                        if (err) {
                                                            console.log('error: ' + err);
                                                            vscode.window.showErrorMessage('error: ' + err);
                                                            return;
                                                        }
                                                        const emptyTempFilePath = stdout.trim();
                                                        var emptyTempPathUri = vscode.Uri.parse("file://" + emptyTempFilePath);
                                                        filesDiff[message.files_diff_index].tempPath = emptyTempPathUri;
                                                        vscode.commands.executeCommand('vscode.diff', emptyTempPathUri, openPath, "SVN DiffTO: " + filesDiffPath, {
                                                            preserveFocus: true,
                                                            viewColumn: vscode.ViewColumn.Beside
                                                        });
                                                    });
                                                }
                                                else {
                                                    var emptyTempPathUri = filesDiff[message.files_diff_index].tempPath;
                                                    vscode.commands.executeCommand('vscode.diff', emptyTempPathUri, openPath, "SVN DiffTO: " + filesDiffPath, {
                                                        preserveFocus: true,
                                                        viewColumn: vscode.ViewColumn.Beside
                                                    });
                                                }
                                            });
                                        });
                                    }
                                    else {
                                        if (typeof filesDiff[message.files_diff_index].tempPath1 === 'undefined') {
                                            const cpTemp1 = require('child_process');
                                            cpTemp1.exec('mktemp', (err, stdout, stderr) => {
                                                if (err) {
                                                    console.log('error: ' + err);
                                                    vscode.window.showErrorMessage('error: ' + err);
                                                    return;
                                                }
                                                const tempFilePath = stdout.trim();
                                                var tempPathUri = vscode.Uri.parse("file://" + tempFilePath);
                                                const svnPath = svnOld + '/' + filesDiff[message.files_diff_index].path;
                                                const cp3 = require('child_process');
                                                cp3.exec('svn cat ' + svnPath + ' > ' + tempFilePath, { maxBuffer: 1024 * 1024 * 1024 }, (err, stdout, stderr) => {
                                                    if (err) {
                                                        console.log('error: ' + err);
                                                        vscode.window.showErrorMessage('error: ' + err);
                                                        return;
                                                    }
                                                    filesDiff[message.files_diff_index].tempPath1 = tempPathUri;
                                                    if (filesDiffType === 'd') {
                                                        const cpTemp2 = require('child_process');
                                                        cpTemp2.exec('mktemp', (err, stdout, stderr) => {
                                                            if (err) {
                                                                console.log('error: ' + err);
                                                                vscode.window.showErrorMessage('error: ' + err);
                                                                return;
                                                            }
                                                            const emptyTempFilePath = stdout.trim();
                                                            var emptyTempPathUri = vscode.Uri.parse("file://" + emptyTempFilePath);
                                                            filesDiff[message.files_diff_index].tempPath2 = emptyTempPathUri;
                                                            vscode.commands.executeCommand('vscode.diff', tempPathUri, emptyTempPathUri, "SVN DiffTO: " + filesDiffPath, {
                                                                preserveFocus: true,
                                                                viewColumn: vscode.ViewColumn.Beside
                                                            });
                                                        });
                                                    }
                                                    else /// m
                                                     {
                                                        const filePath = workspaceFolderPath + '/' + filesDiffPath;
                                                        var openPath = vscode.Uri.parse("file://" + filePath); //A request file path
                                                        filesDiff[message.files_diff_index].tempPath2 = openPath;
                                                        vscode.commands.executeCommand('vscode.diff', tempPathUri, openPath, "SVN DiffTO: " + filesDiffPath, {
                                                            preserveFocus: true,
                                                            viewColumn: vscode.ViewColumn.Beside
                                                        });
                                                    }
                                                });
                                            });
                                        }
                                        else {
                                            vscode.commands.executeCommand('vscode.diff', filesDiff[message.files_diff_index].tempPath1, filesDiff[message.files_diff_index].tempPath2, "SVN DiffTO: " + filesDiffPath, {
                                                preserveFocus: true,
                                                viewColumn: vscode.ViewColumn.Beside
                                            });
                                        }
                                    }
                                    return;
                            }
                        }, undefined, context.subscriptions);
                    });
                });
            });
        });
        quickPick.show();
    });
    context.subscriptions.push(disposable);
}
exports.activate = activate;
function getWebviewContent(difItemsSelectContent, cssUri) {
    return `<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="UTF-8">
		<!--
			Use a content security policy to only allow loading images from https or from our extension directory,
			and only allow scripts that have a specific nonce.
		-->
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		<title>SVN DiffTo</title>
	</head>
	<body height="100%" style="overflow: none;">
		<div>
			<h2>Diff items</h2>
			<select id="diff_select" onchange="onSelectChange()">
				${difItemsSelectContent}
			</select>
		</div>

		<script>
			const vscode = acquireVsCodeApi();
			function escapeHtml(str) {
				return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
			}
			function onSelectChange() {
				var files_diff_index = document.getElementById("diff_select").value;

				vscode.postMessage({
					command: 'onSelectChange',
					files_diff_index: files_diff_index
				})
			}
		</script>
	</body>
</html>`;
}
function getNonce() {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < 32; i++) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map